File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/GoTranslateVisitor.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.language.translator;

import org.antlr.v4.runtime.ParserRuleContext;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;
import org.apache.commons.lang3.StringUtils;
import org.apache.tinkerpop.gremlin.language.grammar.GremlinParser;
import org.apache.tinkerpop.gremlin.process.traversal.strategy.decoration.OptionsStrategy;
import org.apache.tinkerpop.gremlin.structure.VertexProperty;
import org.apache.tinkerpop.gremlin.util.DatetimeHelper;

import java.time.OffsetDateTime;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

public class GoTranslateVisitor extends AbstractTranslateVisitor {
    private final static String GO_PACKAGE_NAME = "gremlingo.";
    private final static List<String> STRATEGY_WITH_MAP_OPTS = Collections.unmodifiableList(Arrays.asList(
            "OptionsStrategy",
            "ReferenceElementStrategy", "ComputerFinalizationStrategy", "ProfileStrategy",
            "ComputerVerificationStrategy", "StandardVerificationStrategy", "VertexProgramRestrictionStrategy"));
    private final static List<String> STRATEGY_WITH_STRING_SLICE = Collections.unmodifiableList(Arrays.asList(
            "ReservedKeysVerificationStrategy", "ProductiveByStrategy"));

    public GoTranslateVisitor() {
        super("g");
    }

    public GoTranslateVisitor(final String graphTraversalSourceName) {
        super(graphTraversalSourceName);
    }

    @Override
    public Void visitDateLiteral(final GremlinParser.DateLiteralContext ctx) {
        // child at 2 is the date argument to datetime() and comes enclosed in quotes
        final String dtString = ctx.getChild(2).getText();
        final OffsetDateTime dt = DatetimeHelper.parse(removeFirstAndLastCharacters(dtString));
        final String zoneInfo = dt.getOffset().getId().equals("Z") ? "UTC+00:00" : "UTC" + dt.getOffset().getId();
        sb.append("time.Date(").append(dt.getYear()).
                append(", ").append(dt.getMonthValue()).
                append(", ").append(dt.getDayOfMonth()).
                append(", ").append(dt.getHour()).
                append(", ").append(dt.getMinute()).
                append(", ").append(dt.getSecond()).
                append(", ").append(dt.getNano()).
                append(", time.FixedZone(\"").append(zoneInfo).append("\", ").append(dt.getOffset().getTotalSeconds()).append(")").
                append(")");
        return null;
    }

    @Override
    public Void visitInfLiteral(final GremlinParser.InfLiteralContext ctx) {
        if (ctx.SignedInfLiteral() != null && ctx.SignedInfLiteral().getText().equals("-Infinity"))
            sb.append("math.Inf(-1)");
        else
            sb.append("math.Inf(1)");
        return null;
    }

    @Override
    public Void visitIntegerLiteral(final GremlinParser.IntegerLiteralContext ctx) {
        String integerLiteral = ctx.getText().toLowerCase();

        // check suffix
        int lastCharIndex = integerLiteral.length() - 1;
        char lastChar = integerLiteral.charAt(lastCharIndex);

        // if the last character is not alphabetic then try to interpret the right type and append the suffix
        if (Character.isAlphabetic(lastChar))
            sb.append(integerLiteral, 0, lastCharIndex);
        else
            sb.append(integerLiteral);
        return null;
    }

    @Override
    public Void visitFloatLiteral(final GremlinParser.FloatLiteralContext ctx) {
        if (ctx.infLiteral() != null) return visit(ctx.infLiteral());
        if (ctx.nanLiteral() != null) return visit(ctx.nanLiteral());

        String floatLiteral = ctx.getText().toLowerCase();

        // check suffix
        int lastCharIndex = floatLiteral.length() - 1;
        char lastChar = floatLiteral.charAt(lastCharIndex);

        // if the last character is not alphabetic then try to interpret the right type and append the suffix
        if (Character.isAlphabetic(lastChar))
            sb.append(floatLiteral, 0, lastCharIndex);
        else
            sb.append(floatLiteral);
        return null;
    }

    @Override
    public Void visitGenericRangeLiteral(final GremlinParser.GenericRangeLiteralContext ctx) {
        throw new TranslatorException("Go does not support range literals");
    }

    @Override
    public Void visitGenericSetLiteral(final GremlinParser.GenericSetLiteralContext ctx) {
        sb.append(GO_PACKAGE_NAME);
        sb.append("NewSimpleSet(");
        for (int i = 0; i < ctx.genericLiteral().size(); i++) {
            final GremlinParser.GenericLiteralContext genericLiteralContext = ctx.genericLiteral(i);
            visit(genericLiteralContext);
            if (i < ctx.genericLiteral().size() - 1)
                sb.append(", ");
        }
        sb.append(")");
        return null;
    }

    @Override
    public Void visitGenericCollectionLiteral(final GremlinParser.GenericCollectionLiteralContext ctx) {
        sb.append("[]interface{}{");
        for (int i = 0; i < ctx.genericLiteral().size(); i++) {
            final GremlinParser.GenericLiteralContext genericLiteralContext = ctx.genericLiteral(i);
            visit(genericLiteralContext);
            if (i < ctx.genericLiteral().size() - 1)
                sb.append(", ");
        }
        sb.append("}");
        return null;
    }

    @Override
    public Void visitGenericMapLiteral(final GremlinParser.GenericMapLiteralContext ctx) {
        sb.append("map[interface{}]interface{}{");
        for (int i = 0; i < ctx.mapEntry().size(); i++) {
            final GremlinParser.MapEntryContext mapEntryContext = ctx.mapEntry(i);
            visit(mapEntryContext);
            if (i < ctx.mapEntry().size() - 1)
                sb.append(", ");
        }
        sb.append(" }");
        return null;
    }

    @Override
    public Void visitMapEntry(final GremlinParser.MapEntryContext ctx) {
        visit(ctx.mapKey());
        sb.append(": ");
        visit(ctx.genericLiteral());
        return null;
    }

    @Override
    public Void visitMapKey(final GremlinParser.MapKeyContext ctx) {
        final int keyIndex = ctx.LPAREN() != null && ctx.RPAREN() != null ? 1 : 0;
        visit(ctx.getChild(keyIndex));
        return null;
    }

    @Override
    public Void visitStringNullableLiteral(GremlinParser.StringNullableLiteralContext ctx) {
        // remove the first and last character (single or double quotes) but only if it is not null
        if (ctx.getText().equals("null")) {
            sb.append("nil");
        } else {
            final String text = removeFirstAndLastCharacters(ctx.getText());
            handleStringLiteralText(text);
        }
        return null;
    }

    @Override
    public Void visitNanLiteral(final GremlinParser.NanLiteralContext ctx) {
        sb.append("math.NaN()");
        return null;
    }

    @Override
    public Void visitNullLiteral(final GremlinParser.NullLiteralContext ctx) {
        sb.append("nil");
        return null;
    }

    @Override
    public Void visitTraversalStrategy(final GremlinParser.TraversalStrategyContext ctx) {
        if (ctx.getChildCount() == 1)
            sb.append(GO_PACKAGE_NAME).append(ctx.getText()).append("()");
        else {
            String strategyName = ctx.getChild(0).getText().equals("new") ? ctx.getChild(1).getText() : ctx.getChild(0).getText();
            sb.append(GO_PACKAGE_NAME).append(strategyName).append("(");
            if (!STRATEGY_WITH_MAP_OPTS.contains(strategyName)) { // omit strategies which use plain map instead of Config struct
                sb.append(GO_PACKAGE_NAME).append(strategyName).append("Config{");
            }

            // get a list of all the arguments to the strategy - i.e. anything not a terminal node
            final List<ParseTree> configs = ctx.children.stream().
                    filter(c -> c instanceof GremlinParser.ConfigurationContext).collect(Collectors.toList());

            // the rest are the arguments to the strategy
            for (int ix = 0; ix < configs.size(); ix++) {
                visit(configs.get(ix));
                if (ix < configs.size() - 1)
                    sb.append(", ");
            }

            if (!Objects.equals(strategyName, "OptionsStrategy")) {
                sb.append("}");
            }
            sb.append(")");
        }
        return null;
    }

    @Override
    public Void visitConfiguration(final GremlinParser.ConfigurationContext ctx) {
        String parent = ctx.getParent().getText();
        String parentName = parent.startsWith("new") ? parent.substring(3, parent.indexOf('(')) : parent.substring(0, parent.indexOf('('));
        if (STRATEGY_WITH_MAP_OPTS.contains(parentName)) { // handle strategies which use plain map instead of Config struct
            sb.append("map[string]interface{}{\"");
            sb.append(ctx.getChild(0).getText());
            sb.append("\": ");
            visit(ctx.getChild(2));
            sb.append("}");
        } else {
            // form of three tokens of key:value to become key=value
            sb.append(SymbolHelper.toGo(ctx.getChild(0).getText()));
            sb.append(": ");
            visit(ctx.getChild(2));
            // handles strategies that takes string slices as config
            if (STRATEGY_WITH_STRING_SLICE.contains(parentName)) {
                final int ix = sb.lastIndexOf("[]interface{}");
                if (ix > 0) {
                    sb.replace(ix, ix +"[]interface{}".length(), "[]string");
                }
            }
        }

        // need to convert List to Set for readPartitions until TINKERPOP-3032
        if (ctx.getChild(0).getText().equals("readPartitions")) {
            final int ix = sb.lastIndexOf("ReadPartitions: [");
            if (ix > 0) {
                final int endIx = sb.indexOf("\"}", ix);
                sb.replace(endIx, endIx + 2, "\")");
                sb.replace(ix, ix + "ReadPartitions: []interface{}{".length(), "ReadPartitions: gremlingo.NewSimpleSet(");
            }

        }

        return null;
    }

    @Override
    public Void visitTraversalSourceSelfMethod_withoutStrategies(final GremlinParser.TraversalSourceSelfMethod_withoutStrategiesContext ctx) {
        sb.append("WithoutStrategies(");
        sb.append(GO_PACKAGE_NAME).append(ctx.classType().getText()).append("()");

        if (ctx.classTypeList() != null && ctx.classTypeList().classTypeExpr() != null) {
            for (GremlinParser.ClassTypeContext classTypeContext : ctx.classTypeList().classTypeExpr().classType()) {
                sb.append(", ").append(GO_PACKAGE_NAME).append(classTypeContext.getText()).append("()");
            }
        }

        sb.append(")");
        return null;
    }

    @Override
    public Void visitUuidLiteral(final GremlinParser.UuidLiteralContext ctx) {
        if (ctx.stringLiteral() == null) {
            sb.append("uuid.New()");
            return null;
        }
        sb.append("uuid.MustParse(");
        visitStringLiteral(ctx.stringLiteral());
        sb.append(")");
        return null;
    }

    @Override
    protected String getCardinalityFunctionClass() {
        return "CardinalityValue";
    }

    protected void visitP(final ParserRuleContext ctx, final Class<?> clazzOfP, final String methodName) {
        sb.append(GO_PACKAGE_NAME);
        super.visitP(ctx, clazzOfP, methodName);
    }

    @Override
    protected String processGremlinSymbol(final String step) {
        return SymbolHelper.toGo(step);
    }

    @Override
    protected void appendExplicitNaming(final String txt, final String prefix) {
        sb.append(GO_PACKAGE_NAME);
        super.appendExplicitNaming(txt, prefix);
    }

    @Override
    protected void appendAnonymousSpawn() {
        sb.append(GO_PACKAGE_NAME).append("T__.");
    }

    static final class SymbolHelper {

        private final static Map<String, String> TO_GO_MAP = new HashMap<>();
        private final static Map<String, String> FROM_GO_MAP = new HashMap<>();

        static {
            TO_GO_MAP.put("OUT", "Out");
            TO_GO_MAP.put("IN", "In");
            TO_GO_MAP.put("BOTH", "Both");
            TO_GO_MAP.put("WithOptions", GO_PACKAGE_NAME + "WithOptions");
            TO_GO_MAP.put("IO", GO_PACKAGE_NAME + "IO");
            TO_GO_MAP.put("__", GO_PACKAGE_NAME + "T__");
            TO_GO_MAP.forEach((k, v) -> FROM_GO_MAP.put(v, k));
        }

        private SymbolHelper() {
            // static methods only, do not instantiate
        }

        public static String toGo(final String symbol) {
            return TO_GO_MAP.getOrDefault(symbol, StringUtils.capitalize(symbol));
        }

        public static String toJava(final String symbol) {
            return FROM_GO_MAP.getOrDefault(symbol, StringUtils.uncapitalize(symbol));
        }

    }
}

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/language/translator/GoTranslateVisitor.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// Context parameters are typed as `any` because the generated grammar files
// require accepting arbitrary context types at runtime.

import TranslateVisitor from './TranslateVisitor.js';
import { TranslatorException } from './TranslatorException.js';

const GO_PACKAGE_NAME = 'gremlingo.';

const STRATEGY_WITH_MAP_OPTS = new Set([
    'OptionsStrategy',
    'ReferenceElementStrategy', 'ComputerFinalizationStrategy', 'ProfileStrategy',
    'ComputerVerificationStrategy', 'StandardVerificationStrategy', 'VertexProgramRestrictionStrategy',
    'MessagePassingReductionStrategy',
]);

const STRATEGY_WITH_STRING_SLICE = new Set([
    'ReservedKeysVerificationStrategy', 'ProductiveByStrategy',
]);

/**
 * Converts a Gremlin traversal string into a Go source code representation.
 * Assumes use of the gremlingo, math, time, and uuid packages.
 */
export default class GoTranslateVisitor extends TranslateVisitor {

    constructor(graphTraversalSourceName: string = 'g') {
        super(graphTraversalSourceName);
    }

    protected override processGremlinSymbol(step: string): string {
        return SymbolHelper.toGo(step);
    }

    protected override getCardinalityFunctionClass(): string {
        return 'CardinalityValue';
    }

    protected override appendExplicitNaming(txt: string, prefix: string): void {
        this.sb.push(GO_PACKAGE_NAME);
        super.appendExplicitNaming(txt, prefix);
    }

    protected override appendAnonymousSpawn(): void {
        this.sb.push(GO_PACKAGE_NAME + 'T__.');
    }

    protected override visitP(ctx: any, pClass: string, methodName: string): void {
        this.sb.push(GO_PACKAGE_NAME);
        super.visitP(ctx, pClass, methodName);
    }

    visitTraversalGType(ctx: any): void {
        const parts: string[] = ctx.getText().split('.');
        this.sb.push(GO_PACKAGE_NAME);
        this.sb.push(this.processGremlinSymbol(parts[0]));
        this.sb.push('.');
        this.sb.push(this.processGremlinSymbol(parts[1].toLowerCase()));
    }

    visitNanLiteral(_ctx: any): void {
        this.sb.push('math.NaN()');
    }

    visitInfLiteral(ctx: any): void {
        if (ctx.SignedInfLiteral() != null && ctx.SignedInfLiteral().getText() === '-Infinity') {
            this.sb.push('math.Inf(-1)');
        } else {
            this.sb.push('math.Inf(1)');
        }
    }

    visitNullLiteral(_ctx: any): void {
        this.sb.push('nil');
    }

    visitStringNullableLiteral(ctx: any): void {
        if (ctx.getText() === 'null') {
            this.sb.push('nil');
        } else {
            const text = TranslateVisitor.removeFirstAndLastCharacters(ctx.getText());
            this.handleStringLiteralText(text);
        }
    }

    visitIntegerLiteral(ctx: any): void {
        const integerLiteral: string = ctx.getText().toLowerCase();
        const lastChar = integerLiteral[integerLiteral.length - 1];
        const num = integerLiteral.slice(0, -1);

        if (!/[a-z]/.test(lastChar)) {
            this.sb.push(integerLiteral);
            return;
        }

        switch (lastChar) {
            case 'b':
                this.sb.push('int8(');
                this.sb.push(num);
                this.sb.push(')');
                break;
            case 's':
                this.sb.push('int16(');
                this.sb.push(num);
                this.sb.push(')');
                break;
            case 'i':
                this.sb.push('int32(');
                this.sb.push(num);
                this.sb.push(')');
                break;
            case 'l':
                this.sb.push('int64(');
                this.sb.push(num);
                this.sb.push(')');
                break;
            case 'n':
                this.sb.push(GO_PACKAGE_NAME + 'ParseBigInt("');
                this.sb.push(num);
                this.sb.push('")');
                break;
            default:
                this.sb.push(integerLiteral);
        }
    }

    visitFloatLiteral(ctx: any): void {
        if (ctx.infLiteral() != null) { this.visit(ctx.infLiteral()); return; }
        if (ctx.nanLiteral() != null) { this.visit(ctx.nanLiteral()); return; }

        const floatLiteral: string = ctx.getText().toLowerCase();
        const lastChar = floatLiteral[floatLiteral.length - 1];
        const num = floatLiteral.slice(0, -1);

        if (!/[a-z]/.test(lastChar)) {
            this.sb.push(floatLiteral);
            return;
        }

        switch (lastChar) {
            case 'f':
                this.sb.push('float32(');
                this.sb.push(num);
                this.sb.push(')');
                break;
            case 'd':
                this.sb.push(num);
                break;
            case 'm':
                this.sb.push(GO_PACKAGE_NAME + 'ParseBigDecimal("');
                this.sb.push(num);
                this.sb.push('")');
                break;
            default:
                this.sb.push(floatLiteral);
        }
    }

    visitDateLiteral(ctx: any): void {
        const dtString: string = ctx.getChild(2).getText();
        const inner = TranslateVisitor.removeFirstAndLastCharacters(dtString);
        const d = new Date(inner);
        if (isNaN(d.getTime())) {
            throw new TranslatorException(`Invalid datetime: ${inner}`);
        }

        const year = d.getUTCFullYear();
        const month = d.getUTCMonth() + 1;
        const day = d.getUTCDate();
        const hours = d.getUTCHours();
        const minutes = d.getUTCMinutes();
        const seconds = d.getUTCSeconds();
        const nanos = d.getUTCMilliseconds() * 1_000_000;
        const totalSeconds = 0; // assume UTC when no offset specified
        const zoneInfo = 'UTC+00:00';

        this.sb.push(`time.Date(${year}, ${month}, ${day}, ${hours}, ${minutes}, ${seconds}, ${nanos}, time.FixedZone("${zoneInfo}", ${totalSeconds}))`);
    }

    visitGenericRangeLiteral(_ctx: any): void {
        throw new TranslatorException('Go does not support range literals');
    }

    visitGenericSetLiteral(ctx: any): void {
        this.sb.push(GO_PACKAGE_NAME + 'NewSimpleSet(');
        const literals: any[] = ctx.genericLiteral();
        for (let i = 0; i < literals.length; i++) {
            this.visit(literals[i]);
            if (i < literals.length - 1) {
                this.sb.push(', ');
            }
        }
        this.sb.push(')');
    }

    visitGenericCollectionLiteral(ctx: any): void {
        this.sb.push('[]interface{}{');
        const literals: any[] = ctx.genericLiteral();
        for (let i = 0; i < literals.length; i++) {
            this.visit(literals[i]);
            if (i < literals.length - 1) {
                this.sb.push(', ');
            }
        }
        this.sb.push('}');
    }

    visitGenericMapLiteral(ctx: any): void {
        this.sb.push('map[interface{}]interface{}{');
        const entries: any[] = ctx.mapEntry();
        for (let i = 0; i < entries.length; i++) {
            this.visit(entries[i]);
            if (i < entries.length - 1) {
                this.sb.push(', ');
            }
        }
        this.sb.push(' }');
    }

    visitMapEntry(ctx: any): void {
        this.visit(ctx.mapKey());
        this.sb.push(': ');
        this.visit(ctx.genericLiteral());
    }

    visitMapKey(ctx: any): void {
        const keyIndex = (ctx.LPAREN() != null && ctx.RPAREN() != null) ? 1 : 0;
        this.visit(ctx.getChild(keyIndex));
    }

    visitUuidLiteral(ctx: any): void {
        if (ctx.stringLiteral() == null) {
            this.sb.push('uuid.New()');
            return;
        }
        this.sb.push('uuid.MustParse(');
        this.visitStringLiteral(ctx.stringLiteral());
        this.sb.push(')');
    }

    visitTraversalStrategy(ctx: any): void {
        if (ctx.getChildCount() === 1) {
            this.sb.push(GO_PACKAGE_NAME);
            this.sb.push(ctx.getText());
            this.sb.push('()');
            return;
        }

        const firstText = ctx.getChild(0).getText();
        const strategyName = firstText === 'new' ? ctx.getChild(1).getText() : firstText;

        this.sb.push(GO_PACKAGE_NAME);
        this.sb.push(strategyName);
        this.sb.push('(');

        if (!STRATEGY_WITH_MAP_OPTS.has(strategyName)) {
            this.sb.push(GO_PACKAGE_NAME);
            this.sb.push(strategyName);
            this.sb.push('Config{');
        }

        const configs: any[] = ctx.configuration();
        for (let ix = 0; ix < configs.length; ix++) {
            this.visit(configs[ix]);
            if (ix < configs.length - 1) {
                this.sb.push(', ');
            }
        }

        if (strategyName !== 'OptionsStrategy') {
            this.sb.push('}');
        }
        this.sb.push(')');
    }

    visitConfiguration(ctx: any): void {
        const parent: string = ctx.parent.getText();
        const parenIdx = parent.indexOf('(');
        const rawName = parenIdx > -1 ? parent.substring(0, parenIdx) : parent;
        const parentName = rawName.startsWith('new') ? rawName.substring(3) : rawName;

        if (STRATEGY_WITH_MAP_OPTS.has(parentName)) {
            this.sb.push('map[string]interface{}{"');
            this.sb.push(ctx.getChild(0).getText());
            this.sb.push('": ');
            this.visit(ctx.getChild(2));
            this.sb.push('}');
            return;
        }

        const key: string = ctx.getChild(0).getText();
        this.sb.push(SymbolHelper.toGo(key));
        this.sb.push(': ');

        const startIdx = this.sb.length;
        this.visit(ctx.genericArgument());

        if (STRATEGY_WITH_STRING_SLICE.has(parentName)) {
            // Replace []interface{} with []string if present
            const idx = this.sb.indexOf('[]interface{}', startIdx);
            if (idx > -1) {
                this.sb[idx] = '[]string';
            }
        }

        if (key === 'readPartitions') {
            // Replace []interface{}{items} with gremlingo.NewSimpleSet(items)
            const added = this.sb.splice(startIdx);
            const joined = added.join('');
            const transformed = joined.replace(/^\[\]interface\{\}\{([\s\S]*)\}$/, GO_PACKAGE_NAME + 'NewSimpleSet($1)');
            this.sb.push(transformed);
        }
    }

    visitTraversalSourceSelfMethod_withoutStrategies(ctx: any): void {
        this.sb.push('WithoutStrategies(');
        this.sb.push(GO_PACKAGE_NAME);
        this.sb.push(ctx.classType().getText());
        this.sb.push('()');

        if (ctx.classTypeList() != null && ctx.classTypeList().classTypeExpr() != null) {
            for (const classTypeCtx of ctx.classTypeList().classTypeExpr().classType()) {
                this.sb.push(', ');
                this.sb.push(GO_PACKAGE_NAME);
                this.sb.push(classTypeCtx.getText());
                this.sb.push('()');
            }
        }

        this.sb.push(')');
    }
}

class SymbolHelper {
    private static readonly TO_GO_MAP: Record<string, string> = {
        'OUT': 'Out',
        'IN': 'In',
        'BOTH': 'Both',
        'bigdecimal': 'BigDecimal',
        'bigint': 'BigInt',
        'datetime': 'DateTime',
        'uuid': 'UUID',
        'vproperty': 'VProperty',
        'WithOptions': GO_PACKAGE_NAME + 'WithOptions',
        'IO': GO_PACKAGE_NAME + 'IO',
        '__': GO_PACKAGE_NAME + 'T__',
    };

    static toGo(symbol: string): string {
        if (symbol in SymbolHelper.TO_GO_MAP) {
            return SymbolHelper.TO_GO_MAP[symbol];
        }
        // Capitalize first letter (mirrors Java StringUtils.capitalize)
        if (!symbol) return symbol;
        return symbol.charAt(0).toUpperCase() + symbol.slice(1);
    }
}
